### PR TITLE
Cromwell backend development docs to readthedocs [BT-353]

### PR DESCRIPTION
**Pull request by mcovarr**
`mkdocs serve` if you want to try this out locally (`brew install mkdocs` may be required).

### CODE REVIEW

File: docs/developers/Backend.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+## Cromwell Backend Development
+
+### What’s a Backend?
+
+Cromwell is a workflow execution service, which means it takes a representation of work to do (tasks) as well as the
+dependencies and code flow between them. When it comes to actually executing the specific task, Cromwell delegates that
+work to a “backend”. A backend is therefore responsible for the actual execution of a single task on some underlying
+computing platform, such as Google Cloud Platform, Azure, AWS, TES, SLURM, etc. Backends are implemented as a software
```
Comment (jgainerdewar) [inline docs/developers/Backend.md:8]: To save everyone future confusion, maybe we shouldn't include Azure and TES both in this list.

File: docs/developers/Backend.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+input file lives in a cloud bucket and is directly copied onto the machine running the command line followed by copying
+the program’s outputs back to a cloud bucket.
+
+The underlying platform can be extremely simple or have arbitrary levels of complexity as long as one can map from the
+above concepts to running a command. That mapping could happen completely in the platform, the Cromwell backend, or a
+mixture of the two. The implementer of a backend will need to strike the balance which works best for both their needs
+and the particulars of the platform itself.
+
+Throughout the rest of this document the following three terms are used, the difference between them is subtle but
+important:
```
Comment (breilly2) [inline docs/developers/Backend.md:39]: Could you add "workflow" for completeness? When reading through the document, the first term related to a unit of work that I came across after this section is "workflow" which made me wonder if I missed something in this section.

File: docs/developers/Backend.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+information to uniquely identify a job. It is recommended that one look in the Cromwell codebase for more information on
+these and other types.
+
+#### BackendWorkflowInitializationActor
+
+If a backend developer wishes to take advantage of the initialization phase of the backend lifecycle they must implement
+this trait. There are three functions which must be implemented:
+
+1. `abortInitialization: Future[WorkflowAbortResponse]` specifies what to do, if anything, when a workflow is requested
+   to abort while a backend initialization is in progress.
```
Comment (breilly2) [inline docs/developers/Backend.md:91]: `BackendWorkflowInitializationActor.scala` says:
```
  def abortInitialization(): Unit = {}
```
I don't see any backends that override this, and it looks like we don't do anything with the result when we call it.

Which one do you think we should change?

File: docs/developers/Backend.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+2. `validate: Future[Unit]` is provided so that the backend can ensure that all of the calls it will handle conform to
+   the rules of that backend. For instance, if a backend requires particular runtime attributes to exist.
+3. `beforeAll: Future[Unit]` is the actual initialization functionality. If a backend requires any work to be done prior
+   to handling a call, that code must be called from here.
```
Comment (breilly2) [inline docs/developers/Backend.md:95]:
```
  def beforeAll(): Future[Option[BackendInitializationData]]
```

File: docs/developers/Backend.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+#### BackendJobExecutionActor
+
+There are three functions to override in this trait
+
+* `execute(jobDescriptor: BackendJobDescriptor): Future[BackendJobExecutionResponse]` requests that the backend run a
```
Comment (jgainerdewar) [inline docs/developers/Backend.md:101]: I think this is out of date - none of these functions take an argument today.

File: docs/developers/Backend.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+  instance type determined from specified runtime attributes (# of CPU, amount of RAM), pull the task’s Docker image,
+  download the input files from Google Cloud Storage onto the VM, run the command in the Docker, copy the output files
+  up to Google Cloud Storage and shut down the VM.
+* `recover(jobDescriptor: BackendJobDescriptor): Future[BackendJobExecutionResponse]` requests that the backend
+  reconnect to a previously launched job. It is recommended that a backend developer implement this as requested but
+  that behavior isn’t appropriate for all situations, in which case it could for instance just call execute. In the case
+  of an SGE backend this might be implemented to determine the SGE ID for the job and start from there.
+* `abortJob(jobKey: BackendJobDescriptorKey): Future[JobAbortResponse]` requests that a running job be aborted. It is
+  recommended that this be properly implemented but that’s not always appropriate. This function could easily be
+  implemented as a no-op.
```
Comment (breilly2) [inline docs/developers/Backend.md:115]: These all look a little off compared to the code. Even the set of functions is a little different. For example, the description for `recover` here seems like it matches the actual `reconnect` function.

**Review by breilly2 — COMMENTED** [review]
Not sure where it would belong, but would it be worth also talking a little bit about `AsyncBackendJobExecutionActor`?

**Comment by jgainerdewar** [comment]
I had the same thought as @breilly2 - we have other classes and traits used in a standard way that people building new backends should consider using. I don't think it makes sense to fully document them here, but it might be helpful to add a note indicating that this universe exists and people may want to explore it. Something like, "Cromwell has a number of classes that extend these traits and implement common backend patterns, which developers may find useful. For example, see..."

File: docs/developers/Backend.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+#### BackendWorkflowFinalizationActor
+
+There are two functions to override for this trait if the backend developer chooses to use the finalization
+functionality
+
+* `abortFinalization: Future[WorkflowAbortResponse]` is called if a workflow is requested to abort while finalization is
+  happening.
```
[the remark this answers is not in the document]
Comment (mcovarr) [reply docs/developers/Backend.md]: yeah there's a lot more skew here than I noticed in my obviously-too-cursory first take on this 🙈

**Comment by jgainerdewar** [comment]
I completed the notes on `StandardAsyncExecutionActor`, this is ready for a final look @breilly2

**Review by jgainerdewar — APPROVED** [review]

**Review by breilly2 — APPROVED** [review]

**Pull request merged**